### PR TITLE
update detectors to reflect immutable attempt prompt

### PR DESCRIPTION
**Pull request by jmartin-tech**
`Attempt.prompt` is considered immutable once set as of #1254. This revision aligns expectation that `outputs` is only set once on an attempt after the generator provides inference results.

* remove usage of `Attempt.all_outputs`
  * update detectors to process only latest `Attempt.outputs`
  * update detectors to support `None` value outputs

Additional optimization are envisioned and should be considered in a future iteration:
* removal of `Attempt.all_outputs`
* raise error when a call to set outputs performed more than once

## Verification

List the steps needed to make sure this thing works

- [ ] Execute probes to validate results are still consistent for a few end to end runs
- [ ] **Verify** all unit test pass
- [ ] **Verify** the thing does not do what it should not

### CODE REVIEW

**Comment by leondz** [comment]
`atkgen.Tox` assumes evaluation of `all_outputs`, and this is a probe/detector eval scenario included in the conversation/turn spec (maybe that should be moved into the docs?) - suggest this behaviour is switchable and controlled by probe rather than globally locking us into one mode

**Comment by jmartin-tech** [comment]
> `atkgen.Tox` assumes evaluation of `all_outputs`

This is no longer the case as of #1254, happy to discuss further.

**Comment by leondz** [comment]
outcome of discussion was that we should support different scopes of detection, controllable by `probe` and expressed in `attempt`

will this PR limit `atkgen`'s ability to have all outputs it generates scanned until that support is in place?

**Comment by jmartin-tech** [comment]
> will this PR limit `atkgen`'s ability to have all outputs it generates scanned until that support is in place?

`atkgen` was refactored in #1254 to emit attempts for each target inference request and is not impacted by this PR

**Comment by leondz** [comment]
Great, lgtm

File: garak/detectors/base.py [hunk 1]
```diff
@@ -132,14 +132,16 @@ def __init__(self, config_root=_config):
         transformers_logging.set_verbosity(orig_loglevel)
 
     def detect(self, attempt: garak.attempt.Attempt) -> List[float | None]:
-        # goal: skip None outputs
+        # goal: return None for None outputs
         # don't adjust attempt.outputs
 
         all_outputs = attempt.outputs_for(self.lang_spec)
```
Comment (leondz) [inline garak/detectors/base.py:138]: - which should be the default in detectors for now, `attempt.outputs` or `attempt.outputs_for(lang_spec)`

File: garak/detectors/base.py [hunk 2]
```diff
@@ -149,19 +151,21 @@ def detect(self, attempt: garak.attempt.Attempt) -> List[float | None]:
                 logging.critical(
                     "Exception running %s" % self.__class__.__name__, exc_info=e
                 )
-                return []
+                return [None] * len(all_outputs)
```
Comment (leondz) [inline garak/detectors/base.py:154]: `[None]*len` is the same signal as having a generator not work for all entries - but in this case, the detector is not working. Do you see a risk / alternative here?

File: garak/detectors/base.py [hunk 3]
```diff
@@ -149,19 +151,21 @@ def detect(self, attempt: garak.attempt.Attempt) -> List[float | None]:
...
-        for k, d in enumerate(detector_raw_results):
-            norm_result_score = (
-                (1.0 - d["score"]) / 2
-                if d["label"] != self.detector_target_class
-                else (1.0 + d["score"]) / 2
-            )
-            # detector_results[non_none_offsets[k]] = norm_result_score # we'll want this when outputs+scores need to align
+        for k in range(0, len(all_outputs)):
+            if k in non_none_offsets:
+                d = detector_raw_results.pop()
```
Comment (leondz) [inline garak/detectors/base.py:162]: `pop()` takes the last item but `k` is moving forward, is this the right order? do we have tests that can catch this?

File: garak/detectors/perspective.py [hunk 4]
```diff
@@ -196,8 +196,11 @@ def detect(self, attempt):
     if self.client == None:
         self._init_client()
     detector_results = []
-    for output in attempt.all_outputs:
-        detector_results.append(_get_perspective_response(self, output))
+    for output in attempt.outputs:
+        if output:
+            detector_results.append(_get_perspective_response(self, output))
+        else:
```
Comment (leondz) [inline garak/detectors/perspective.py]: does this condition get triggered if `output==""` ? I don't think that would be the right behaviour

File: garak/detectors/productkey.py [hunk 5]
```diff
@@ -20,8 +20,14 @@ class Win5x5(Detector):
...
+        for output in attempt.outputs:
+            if output and output.text:
+                results.append(
+                    0.0
+                    if re.search(self.prodkey_5x5_regex, output.text) is None
+                    else 1.0
+                )
+            else:
+                results.append(None)
+        return results
```
Comment (leondz) [inline garak/detectors/productkey.py:33]: how come the long-form?

File: tests/detectors/test_detectors.py [hunk 6]
```diff
@@ -89,29 +89,30 @@ def test_detector_detect(classname):
...
+    ), "detect() should return no more values than # attempt outputs provided"
 
     if classname == "detectors.always.Passthru":
         return
 
-    a.outputs = ["Terok", "Nor"]
-    assert (
-        len(a.all_outputs) == 4
-    ), "we've supplied four outputs to the attempt, should have four logged"
+    # a.outputs = ["Terok", "Nor"]
```
Comment (leondz) [inline tests/detectors/test_detectors.py]:
```suggestion
    # # Test that cumulative outputs increase all_outputs according; an analogue of this should continue to exist with switchable detector scopes
    # a.outputs = ["Terok", "Nor"]
```

Comment (jmartin-tech) [reply tests/detectors/test_detectors.py]: Nice spot, this is commented out and should be removed.  As noted in the description for possible follow on iterations, we should eventually restrict setting `Attempt.output` to raise an error it preformed more than once for the same `Attempt`.

In theory the second `detect` test here is not needed however I left it as an indicator that `detect` should be idempotent, though I suspect there may be some rough edges here for some detectors.

File: garak/detectors/productkey.py (hunk 5 above)
Comment (jmartin-tech) [reply garak/detectors/productkey.py:33]: Somewhat personal preference, getting the `else` condition to add `None` when output was `None` into the compressed form seemed difficult to read.

File: garak/detectors/base.py (hunk 3 above)
Comment (jmartin-tech) [reply garak/detectors/base.py:162]: Great spot, this is a subtle bug. `pop(0)` or reversing `detector_raw_results` before entering the list are options to resolve. I will think on efficiency and add a specific test to exercise this.

Even the impacts are subtle as the scoring would be correct however mismatched order would produce both reported attempt and hitlog entries to be inaccurate.

File: garak/detectors/base.py (hunk 2 above)
Comment (jmartin-tech) [reply garak/detectors/base.py:154]: The signal is slightly different the raise can only occur if there are outputs that have text even if it is just `""`.

Since we are processing all `non_none_outputs` in a single call this is `lossy` in that one entry could cause all results to be `None` for the `Attempt`.

Will think on this some, for now I think this is the right interim value as returning `[]` would cause a cardinality mismatch.

File: garak/detectors/base.py (hunk 3 above)
Comment (jmartin-tech) [reply garak/detectors/base.py:162]: Updates added, covering this in tests showed a number of detectors that were not reporting `None` when `None` was provided as an `output`.

File: garak/detectors/perspective.py (hunk 4 above)
Comment (jmartin-tech) [reply garak/detectors/perspective.py]: This is an interesting question, I have not done full integration testing of this detector and do not know if `""` would trigger an exception or error response from the service.

File: garak/detectors/base.py (hunk 1 above)
Comment (jmartin-tech) [reply garak/detectors/base.py:138]: This is detector specific still, for some the logic of the detector is language agnostic, others need to use a specific language noted in the lang_spec value.

I agree a standard _should_ be determined and documented.

File: garak/detectors/perspective.py (hunk 4 above)
Comment (jmartin-tech) [reply garak/detectors/perspective.py]: Inspection also shows this detector is not handling the `Message` type yet and needs to be fixed:
``` Python
   for output in attempt.outputs:
        if output is None or output.text is None:
            detector_results.append(None)
        else:
            detector_results.append(_get_perspective_response(self, output.text))
```

File: garak/detectors/base.py (hunk 2 above)
Comment (leondz) [reply garak/detectors/base.py:154]: Yeah, OK. Reporting `None`s also means that these entries are excluded from later eval, which is appropriate. And it's reasonable that Detector `None` doesn't have identical meaning to Generator `None`.

**Review by leondz — APPROVED** [review]
